File: src/test/java/com/koddy/server/member/domain/model/mentee/MenteeTest.java
package com.koddy.server.member.domain.model.mentee;

import com.koddy.server.common.UnitTest;
import com.koddy.server.common.fixture.LanguageFixture;
import com.koddy.server.member.domain.model.Language;
import com.koddy.server.member.exception.MemberException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;

import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_1;
import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_2;
import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_3;
import static com.koddy.server.member.domain.model.Member.Status.ACTIVE;
import static com.koddy.server.member.domain.model.Role.MENTEE;
import static com.koddy.server.member.exception.MemberExceptionCode.MAIN_LANGUAGE_MUST_BE_ONLY_ONE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;

@DisplayName("Member/Mentee -> 도메인 Aggregate [Mentee] 테스트")
class MenteeTest extends UnitTest {
    @Nested
    @DisplayName("초기 Mentee 생성")
    class Construct {
        @Test
        @DisplayName("사용 가능한 메인 언어가 1개가 아니면(0 or N) 정책에 의거해서 Mentee를 생성할 수 없다")
        void throwExceptionByMainLanguageIsNotOne() {
            // given
            final List<Language> languages = List.of(
                    LanguageFixture.KR_MAIN.toDomain(),
                    LanguageFixture.EN_MAIN.toDomain(),
                    LanguageFixture.JP_SUB.toDomain()
            );

            // when - then
            assertThatThrownBy(() -> new Mentee(
                    MENTEE_1.getPlatform(),
                    MENTEE_1.name(),
                    MENTEE_1.getProfileImageUrl(),
                    MENTEE_1.getNationality(),
                    languages,
                    MENTEE_1.getInterest()
            ))
                    .isInstanceOf(MemberException.class)
                    .hasMessage(MAIN_LANGUAGE_MUST_BE_ONLY_ONE.getMessage());
        }

        @Test
        @DisplayName("초기 Mentee를 생성한다")
        void success() {
            final Mentee mentee = new Mentee(
                    MENTEE_1.getPlatform(),
                    MENTEE_1.getName(),
                    MENTEE_1.getProfileImageUrl(),
                    MENTEE_1.getNationality(),
                    MENTEE_1.getLanguages(),
                    MENTEE_1.getInterest()
            );

            assertAll(
                    // Required
                    () -> assertThat(mentee.getPlatform().getProvider()).isEqualTo(MENTEE_1.getPlatform().getProvider()),
                    () -> assertThat(mentee.getPlatform().getSocialId()).isEqualTo(MENTEE_1.getPlatform().getSocialId()),
                    () -> assertThat(mentee.getPlatform().getEmail().getValue()).isEqualTo(MENTEE_1.getPlatform().getEmail().getValue()),
                    () -> assertThat(mentee.getName()).isEqualTo(MENTEE_1.getName()),
                    () -> assertThat(mentee.getNationality()).isEqualTo(MENTEE_1.getNationality()),
                    () -> assertThat(mentee.getProfileImageUrl()).isEqualTo(MENTEE_1.getProfileImageUrl()),
                    () -> assertThat(mentee.getStatus()).isEqualTo(ACTIVE),
                    () -> assertThat(mentee.getLanguages()).containsExactlyInAnyOrderElementsOf(MENTEE_1.getLanguages()),
                    () -> assertThat(mentee.getRole()).isEqualTo(MENTEE),
                    () -> assertThat(mentee.getInterest().getSchool()).isEqualTo(MENTEE_1.getInterest().getSchool()),
                    () -> assertThat(mentee.getInterest().getMajor()).isEqualTo(MENTEE_1.getInterest().getMajor()),

                    // Optional
                    () -> assertThat(mentee.getIntroduction()).isNull(),

                    // isCompleted
                    () -> assertThat(mentee.isProfileComplete()).isFalse()
            );
        }
    }

    @Test
    @DisplayName("Mentee 프로필이 완성되었는지 확인한다 (자기소개)")
    void isProfileComplete() {
        /* menteeA 완성 */
        final Mentee menteeA = MENTEE_1.toDomain();
        assertThat(menteeA.isProfileComplete()).isTrue();

        /* menteeB 1차 회원가입 */
        final Mentee menteeB = new Mentee(
                MENTEE_2.getPlatform(),
                MENTEE_2.getName(),
                MENTEE_2.getProfileImageUrl(),
                MENTEE_2.getNationality(),
                MENTEE_2.getLanguages(),
                MENTEE_2.getInterest()
        );
        assertThat(menteeB.isProfileComplete()).isFalse();

        /* menteeB 프로필 완성 */
        menteeB.completeInfo(MENTEE_2.getIntroduction());
        assertThat(menteeB.isProfileComplete()).isTrue();

        /* menteeC 완성 */
        final Mentee menteeC = MENTEE_3.toDomain();
        assertThat(menteeC.isProfileComplete()).isTrue();

        /* menteeC 미완성 진행 */
        menteeC.completeInfo(null);
        assertThat(menteeC.isProfileComplete()).isFalse();
    }

    @Nested
    @DisplayName("Mentee 기본 정보 수정")
    class UpdateBasicInfo {
        @Test
        @DisplayName("Mentee 기본 정보를 수정한다")
        void success() {
            // given
            final Mentee mentee = MENTEE_1.toDomain().apply(1L);

            // when
            mentee.updateBasicInfo(
                    MENTEE_2.getName(),
                    MENTEE_2.getNationality(),
                    MENTEE_2.getProfileImageUrl(),
                    MENTEE_2.getIntroduction(),
                    MENTEE_2.getLanguages(),
                    MENTEE_2.getInterest().getSchool(),
                    MENTEE_2.getInterest().getMajor()
            );

            // then
            assertAll(
                    () -> assertThat(mentee.getName()).isEqualTo(MENTEE_2.getName()),
                    () -> assertThat(mentee.getNationality()).isEqualTo(MENTEE_2.getNationality()),
                    () -> assertThat(mentee.getProfileImageUrl()).isEqualTo(MENTEE_2.getProfileImageUrl()),
                    () -> assertThat(mentee.getIntroduction()).isEqualTo(MENTEE_2.getIntroduction()),
                    () -> assertThat(mentee.getLanguages()).containsExactlyInAnyOrderElementsOf(MENTEE_2.getLanguages()),
                    () -> assertThat(mentee.getInterest().getSchool()).isEqualTo(MENTEE_2.getInterest().getSchool()),
                    () -> assertThat(mentee.getInterest().getMajor()).isEqualTo(MENTEE_2.getInterest().getMajor())
            );
        }
    }
}

File: src/test/java/com/koddy/server/member/domain/model/mentee/MenteeTest.kt
package com.koddy.server.member.domain.model.mentee

import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.fixture.LanguageFixture
import com.koddy.server.common.fixture.MenteeFixtureStore.menteeFixture
import com.koddy.server.member.domain.model.Member
import com.koddy.server.member.domain.model.Role
import com.koddy.server.member.exception.MemberException
import com.koddy.server.member.exception.MemberExceptionCode.AVAILABLE_LANGUAGE_MUST_EXISTS
import com.koddy.server.member.exception.MemberExceptionCode.MAIN_LANGUAGE_MUST_BE_ONLY_ONE
import io.kotest.assertions.assertSoftly
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.FeatureSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import io.kotest.matchers.throwable.shouldHaveMessage

@UnitTestKt
@DisplayName("Member/Mentee -> 도메인 Aggregate [Mentee] 테스트")
internal class MenteeTest : FeatureSpec({
    feature("Mentee 생성") {
        val fixture = menteeFixture(sequence = 1)

        scenario("메인 언어를 (0..N)개 선택했다면 Mentee를 생성할 수 없다") {
            shouldThrow<MemberException> {
                Mentee(
                    platform = fixture.platform,
                    name = fixture.name,
                    nationality = fixture.nationality,
                    languages = listOf(),
                    interest = fixture.interest,
                )
            } shouldHaveMessage AVAILABLE_LANGUAGE_MUST_EXISTS.message

            shouldThrow<MemberException> {
                Mentee(
                    platform = fixture.platform,
                    name = fixture.name,
                    nationality = fixture.nationality,
                    languages = listOf(
                        LanguageFixture.KR_MAIN.toDomain(),
                        LanguageFixture.EN_MAIN.toDomain(),
                        LanguageFixture.JP_SUB.toDomain(),
                    ),
                    interest = fixture.interest,
                )
            } shouldHaveMessage MAIN_LANGUAGE_MUST_BE_ONLY_ONE.message
        }

        scenario("메인 언어를 정확히 1개 선택했다면 Mentee를 생성할 수 있다") {
            listOf(
                listOf(LanguageFixture.KR_MAIN.toDomain()),
                listOf(
                    LanguageFixture.KR_MAIN.toDomain(),
                    LanguageFixture.EN_SUB.toDomain(),
                    LanguageFixture.JP_SUB.toDomain(),
                ),
            ).forEach { language ->
                val result = Mentee(
                    platform = fixture.platform,
                    name = fixture.name,
                    nationality = fixture.nationality,
                    languages = language,
                    interest = fixture.interest,
                )

                assertSoftly {
                    // Common
                    result.platform.provider shouldBe fixture.platform.provider
                    result.platform.socialId shouldBe fixture.platform.socialId
                    result.platform.email!!.value shouldBe fixture.platform.email!!.value
                    result.name shouldBe fixture.name
                    result.nationality shouldBe fixture.nationality
                    result.introduction shouldBe null
                    result.profileImageUrl shouldBe null
                    result.profileComplete shouldBe false
                    result.status shouldBe Member.Status.ACTIVE
                    result.role shouldBe Role.MENTEE
                    result.languages.map { it.category } shouldContainExactly language.map { it.category }
                    result.languages.map { it.type } shouldContainExactly language.map { it.type }

                    // Mentee
                    result.interest.school shouldBe fixture.interest.school
                    result.interest.major shouldBe fixture.interest.major
                }
            }
        }
    }

    feature("Mentee's completeProfile & profileComplete") {
        val fixtureA = menteeFixture(id = 1L)
        val fixtureB = menteeFixture(id = 2L)
        val fixtureC = menteeFixture(id = 3L)

        scenario("Mentee 프로필이 완성되었는지 확인한다 [자기소개 & 프로필 이미지]") {
            /**
             * MenteeA
             */
            // 완성
            val menteeA: Mentee = fixtureA.toDomain()
            menteeA.profileComplete shouldBe true

            /**
             * MenteeB
             */
            // 미완성
            val menteeB = Mentee(
                id = fixtureB.id,
                platform = fixtureB.platform,
                name = fixtureB.name,
                nationality = fixtureB.nationality,
                languages = fixtureB.languages,
                interest = fixtureB.interest,
            )
            menteeB.profileComplete shouldBe false

            // 완성
            menteeB.completeProfile(
                introduction = fixtureB.introduction,
                profileImageUrl = fixtureB.profileImageUrl,
            )
            menteeB.profileComplete shouldBe true

            /**
             * MenteeC
             */
            // 완성
            val menteeC: Mentee = fixtureC.toDomain()
            menteeC.profileComplete shouldBe true

            // 미완성
            menteeC.completeProfile(
                introduction = null,
                profileImageUrl = fixtureC.profileImageUrl,
            )
            menteeC.profileComplete shouldBe false
        }
    }

    feature("Mentee's updateBasicInfo") {
        val fixtureA = menteeFixture(sequence = 1)
        val fixtureB = menteeFixture(sequence = 2)

        scenario("Mentee 기본 정보를 수정한다") {
            val mentee: Mentee = fixtureA.toDomain()

            mentee.updateBasicInfo(
                name = fixtureB.name,
                nationality = fixtureB.nationality,
                profileImageUrl = fixtureB.profileImageUrl,
                introduction = fixtureB.introduction,
                languages = fixtureB.languages,
                interestSchool = fixtureB.interest.school,
                interestMajor = fixtureB.interest.major,
            )
            assertSoftly {
                // Effected
                mentee.name shouldBe fixtureB.name
                mentee.nationality shouldBe fixtureB.nationality
                mentee.introduction shouldBe fixtureB.introduction
                mentee.profileImageUrl shouldBe fixtureB.profileImageUrl
                mentee.languages.map { it.category } shouldContainExactly fixtureB.languages.map { it.category }
                mentee.languages.map { it.type } shouldContainExactly fixtureB.languages.map { it.type }
                mentee.interest.school shouldBe fixtureB.interest.school
                mentee.interest.major shouldBe fixtureB.interest.major

                // Not Effected
                mentee.platform.provider shouldBe fixtureA.platform.provider
                mentee.platform.socialId shouldBe fixtureA.platform.socialId
                mentee.platform.email!!.value shouldBe fixtureA.platform.email!!.value
                mentee.profileComplete shouldBe true
                mentee.status shouldBe Member.Status.ACTIVE
                mentee.role shouldBe Role.MENTEE
            }
        }
    }
})
